api/apiroutes.go

```diff
@@ -1299,6 +1303,48 @@ func (c *appContext) addressTotals(w http.ResponseWriter, r *http.Request) {
 	writeJSON(w, totals, c.getIndentQuery(r))
 }
 
+// For /download/address/io/{address}
+func (c *appContext) addressIoCsv(w http.ResponseWriter, r *http.Request) {
+	if c.LiteMode {
+		http.Error(w, http.StatusText(http.StatusNotImplemented), http.StatusNotImplemented)
+		return
+	}
+
+	address := m.GetAddressCtx(r)
+	if address == "" {
+		log.Debugf("Failed to parse address from request")
+		http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+		return
+	}
+
+	_, _, addrErr := txhelpers.AddressValidation(address, c.Params)
+	if addrErr != nil {
+		log.Debugf("Error validating address %s: %v", address, addrErr)
+		http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
+		return
+	}
+
+	w.Header().Set("Content-Disposition",
+		fmt.Sprintf("attachment;filename=address-io-%d-%s.csv",
+			c.AuxDataSource.Height(), fmt.Sprintf("%x", time.Now().Unix())))
```

Comment:
Related, should there be a `c.statusMtx.RLock()` [here](https://github.com/decred/dcrdata/blob/f050e3befaee23bc673c115abe0dae1fd19c80d2/api/apiroutes.go#L285)?

Reply:
Mmmm, yes I think so.  The `apitypes.Status` type could actually use it's own embedded mutex with `Height()` accessor to wrap the access with lock/unlock, or adding a mutex to the Status type gets too ugly for some reason an Height method for appContext.

Reply:
It might have gotten ugly. The apitypes have json tags, so presumably should be able to be encoded/decoded to/from  json, during which time the `height` member would be accessed directly. It also looks like there's a function `m.StatusCtx` to embed the status in a request context which might muddy up the waters, but it doesn't look like any endpoints are using that right now.

Reply:
I was thinking along the lines of:

```go
// Status indicates the state of the server, including the API version and the
// software version.
type Status struct {
	sync.RWMutex    `json:"-"`
	Ready           bool   `json:"ready"`
	DBHeight        uint32 `json:"db_height"`
	DBLastBlockTime int64  `json:"db_block_time"`
	Height          uint32 `json:"node_height"`
	NodeConnections int64  `json:"node_connections"`
	APIVersion      int    `json:"api_version"`
	DcrdataVersion  string `json:"dcrdata_version"`
	NetworkName     string `json:"network_name"`
}

func (s *Status) GetHeight() uint32 {
	s.RLock()
	defer s.RUnlock()
	return s.Height
}

func (s *Status) SetHeight(height uint32) {
	s.Lock()
	defer s.Unlock()
	s.Height = height
}
```

And when it needs to be marshalled,

```go
func (c *appContext) status(w http.ResponseWriter, r *http.Request) {
	c.Status.RLock()
	defer c.Status.RUnlock()
	writeJSON(w, &c.Status, c.getIndentQuery(r))
}
```

Reply:
And `appContext` can lose it's `statusMtx` field as long as everywhere that stuff like  `c.statusMtx.Lock()` happens is replaced with `c.Status.Lock()`